library/general/src/lib/cfa/nsswitch.rb

```diff
@@ -0,0 +1,273 @@
+    def entries
+      databases.keys
+    end
+
+    # Service specifications for given database name
+    #
+    # @param db_name [String] the database name, e.g., "passwd" or "hosts"
+    #
+    # @return [Array<String>] database service specifications
+    def services_for(db_name)
+      databases[db_name]&.services
+    end
+
+    # Update (or create if it does not exist yet) the entry for given database name
+    #
+    #
+    # @note Current services, if any, are completely replaced
+    #
+    # @see #entry_for
+    # @see DatabaseEntry#services=
+    #
+    # @param db_name [String] the database name, e.g., "passwd" or "hosts"
+    # @param services [Array<String>] service specifications for the database
+    def update_entry(db_name, services)
+      database = entry_for(db_name)
+      database.services = services
+    end
+
+    # Delete entry for given database name
+    #
+    # @param db_name [String] the database db_name, e.g., "passwd" or "hosts"
+    def delete_entry(db_name)
+      data.delete(database_matcher(db_name))
+    end
+
+    # Writes the current content to /etc/nsswitch.conf
+    #
+    # @see CFA::BaseModel#save
+    def save
+      return false unless modified?
+
+      new_content = @parser.serialize(data)
+
+      @parser.file_name = write_path if @parser.respond_to?(:file_name=)
+      @file_handler.write(write_path, new_content)
+      @current_content = new_content
+
+      true
+    end
+
+    # Whether the content has changed
+    #
+    # @return [Boolean] true if content has been modified; false otherwise.
+    def modified?
+      @current_content != @parser.serialize(data)
+    end
+
+  private
+
+    # The path for loading the content
+    #
+    # @return [String] path to local configuration file if it exists; path to vendor one otherwise
+    def load_path
+      Yast::FileUtils.Exists(LOCAL_PATH) ? LOCAL_PATH : VENDOR_PATH
+    end
+
+    # The path for writing the content
+    #
+    # @return [String] path to local configuration file
+    def write_path
+      LOCAL_PATH
+    end
+
+    # Return the CFA::Matcher for given database name
+    #
+    # @param db_name [String] the database name, e.g., "passwd" or "hosts"
+    #
+    # @return [CFA::Matcher]
+    def database_matcher(db_name)
+      CFA::Matcher.new { |k, v| k == "database[]" && v.value == db_name }
+    end
+
+    # Find or create the DatabaseEntry for given database name
+    #
+    # @param db_name [String] the database name, e.g., "passwd" or "hosts"
+    #
+    # @return [DatabaseEntry]
+    def entry_for(db_name)
+      return databases[db_name] if databases.keys.include?(db_name)
+
+      database = CFA::AugeasTreeValue.new(CFA::AugeasTree.new, db_name)
+      data.add("database[]", database)
+      DatabaseEntry.new(database)
+    end
+
+    # Available databases
+    #
+    # @return [Hash{String => DatabaseEntry}]
+    def databases
+      raw_databases.each_with_object({}) do |database, collection|
```

Comment:
Do you measure performance in some real scenarios? As each call to databases recreates whole hash, so a quite performance suboptimal.

Reply:
I was worried about this too, but I didn't notice a huge penalty when working with it in an `irb` console using my own /usr/etc/nsswitch.conf file.

For sure, it would be nice to optimize it, do you have any suggestion/idea to do so?

Reply:
well, you can cache it and when doing operations that modify database then clear cache

Reply:
To be honest, I discarded this idea because I thought it does not worth it. Let's finish polishing the other aspects first and postpone this for the end, if you don't mind.